src/lib/storage/table.cpp

```diff
@@ -18,75 +18,68 @@
 
 namespace opossum {
 
-Table::Table(const ChunkOffset target_chunk_size) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
-}
+Table::Table(const ChunkOffset target_chunk_size) : _target_chunk_size{target_chunk_size} { create_new_chunk(); }
 
 void Table::add_column(const std::string& name, const std::string& type) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  _column_names.push_back(name);
+  _column_types.push_back(type);
+  for (const auto& chunk : _chunks) {
+    resolve_data_type(type, [&](const auto data_type_t) {
+      using ColumnDataType = typename decltype(data_type_t)::type;
+      chunk->add_segment(std::make_shared<ValueSegment<ColumnDataType>>());
+    });
+  }
 }
 
 void Table::append(const std::vector<AllTypeVariant>& values) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  // if current chunk size is maxed out, create a new chunk, adding as many segments as we have columns
+  if (_chunks.back()->size() >= target_chunk_size()) {
+    create_new_chunk();
+  }
+
+  _chunks.back()->append(values);
 }
 
 void Table::create_new_chunk() {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  auto new_chunk = std::make_shared<Chunk>();
+  for (const auto& column_type : _column_types) {
+    resolve_data_type(column_type, [&](const auto data_type_t) {
+      using ColumnDataType = typename decltype(data_type_t)::type;
+      new_chunk->add_segment(std::make_shared<ValueSegment<ColumnDataType>>());
+    });
+  }
+  _chunks.push_back(new_chunk);
 }
 
-ColumnCount Table::column_count() const {
-  // Implementation goes here
-  return ColumnCount{0};
-}
+ColumnCount Table::column_count() const { return static_cast<ColumnCount>(_column_names.size()); }
 
 ChunkOffset Table::row_count() const {
-  // Implementation goes here
-  return 0;
+  auto total_rows = ChunkOffset{0};
+  for (const auto& chunk : _chunks) {
+    total_rows += chunk->size();
+  }
+  return total_rows;
 }
 
-ChunkID Table::chunk_count() const {
-  // Implementation goes here
-  return ChunkID{0};
-}
+ChunkID Table::chunk_count() const { return static_cast<ChunkID>(_chunks.size()); }
 
 ColumnID Table::column_id_by_name(const std::string& column_name) const {
-  // Implementation goes here
-  return ColumnID{0};
+  auto column_name_location = std::find(column_names().begin(), column_names().end(), column_name);
+  Assert(column_name_location != column_names().end(), "The given column name is not contained in the table.");
+  return ColumnID{static_cast<ColumnID>(std::distance(column_names().begin(), column_name_location))};
```

Comment:
We would recommend using one type of cast.

Reply:
With the present static_cast, constructing an additional Column object (the outer `ColumnID{...}` is unnecessary.